[visionOS] Fix the drag preview of the <model> element

achan00 opened the pull request:
#### 676b1c75b77ea731918213ca6853ff55aaccec61
<pre>
[visionOS] Fix the drag preview of the &lt;model&gt; element
<a href="https://bugs.webkit.org/show_bug.cgi?id=287525">https://bugs.webkit.org/show_bug.cgi?id=287525</a>
<a href="https://rdar.apple.com/141165213">rdar://141165213</a>

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

For UITargetedDragPreview to capture the snapshot of a view
with entity-based content properly, we need to pass the _UIRemoteView
hosting the model content directly when constructing the UITargetedDragPreview.
To do that, we need to pass the model element’s PlatformLayerIdentifier
in the DragItem, so we can use it to retrieve the _UIRemoteView hosting
the model entity from ModelPresentationManagerProxy in the UI process
to stage the drag session with.
In DragDropInteractionState::createDragPreviewInternal(), if the drag
preview content is the _UIRemoteView rather than an image, create the
UITargetedDragPreview with that view directly.

Handle the edge case where the model element is removed while
it’s being dragged by adding it to the WKContentView’s drag previews
container view.

* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::doSystemDrag):
If the dragged element is an HTMLModelElement, save off its
PlatformLayerIdentifier in the DragItem.
* Source/WebCore/platform/DragItem.h:
Include the model element’s PlatformLayerIdentifier
in the DragItem struct.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _willInvalidateDraggedModelWithContainerView:]):
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.mm:
(WebKit::ModelPresentationManagerProxy::startDragForModel):
Add the dragged model’s layer identifier to m_activelyDraggedModelLayerIDs.
Return the _UIRemoteView hosting the model entity.
(WebKit::ModelPresentationManagerProxy::doneWithCurrentDragSession):
Clear m_activelyDraggedModelLayerIDs.
(WebKit::ModelPresentationManagerProxy::invalidateModel):
If the model layer being removed is currently being dragged, call
-[WKWebView _willInvalidateDraggedModelWithContainerView:] with the
model container view so it can be re-parented to the window’s
view hierarchy.
* Source/WebKit/UIProcess/ios/DragDropInteractionState.h:
DragSourceState now stores content with the DragPreviewContentType
which is either a UIImage or a UIView.
* Source/WebKit/UIProcess/ios/DragDropInteractionState.mm:
(WebKit::shouldUseDragImageToCreatePreviewForDragSource):
(WebKit::DragDropInteractionState::previewForCancelling):
The UITargetedDragPreview’s view is the _UIRemoteView’s that’s
already in the remote layer tree’s view hierarchy if the dragged
element is a model element. Only add UITargetedDragPreview’s view
to m_previewViewsForDragCancel if its superview is the preview container.
(WebKit::DragDropInteractionState::createDragPreviewInternal const):
If the drag preview content is a UIView (which is the case for a
dragged model), create a UITargetedDragPreview with that UIView.
(WebKit::DragDropInteractionState::stageDragItem):
(WebKit::DragDropInteractionState::updatePreviewsForActiveDragSources):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _startDrag:item:]):
If the dragged item has a model layer identifier, stage the drag
item with the _UIRemoteView returned from the page’s ModelPresentationManagerProxy.
(-[WKContentView cleanUpDragSourceSessionState]):
Clear the dragged models tracked in ModelPresentationManagerProxy.
(-[WKContentView _willInvalidateDraggedModelWithContainerView:]):
This is called when the dragged model is being removed from the document.
We need to keep its container view in the window by hiding it and
adding it to the WKContentView’s _dragPreviewContainerView.
* Source/WebKit/UIProcess/ios/WKPageHostedModelView.mm:
(-[WKPageHostedModelView init]):
Add some nil checks which are needed for TestWebKitAPI model drag
and drop tests.
* Tools/TestRunnerShared/spi/UIKitSPIForTesting.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/simple-model-page.html: Added.
Simple test page with a model. Posts messages when the model is
loaded and when it’s ready.
* Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
(-[ModelLoadingMessageHandler userContentController:didReceiveScriptMessage:]):
Teach ModelLoadingMessageHandler to catch both the model loaded
event and the model ready event.
(TestWebKitAPI::TEST(DragAndDropTests, CanStartDragOnModel)):
Update test to load the new simple-model-page.html.
(TestWebKitAPI::TEST(DragAndDropTests, CheckModelDragPreview)):
Simulate the dragging of a model element and verify that the resulting
UITargetedDragPreview’s view is a _UIRemoteView. In the middle of the
drag, remove the model from the document and make sure the _UIRemoteView
is still in the window’s view hierarchy. We have to test this scenario
while the drag is still in progress because the page’s drag preview
container view gets removed from the window at the end of the drag session.
* Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:
* Tools/TestWebKitAPI/ios/DragAndDropSimulatorIOS.mm:
(-[DragAndDropSimulator _advanceProgress]):
Call _dragProgressMidPointReachedBlock if it&apos;s set when the
drag progress has passed the midway point.

Canonical link: <a href="https://commits.webkit.org/290437@main">https://commits.webkit.org/290437@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d504b3db9e00e70fbd6308a548717d6771aaa48d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9515 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44885 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94986 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40759 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/92038 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17864 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69280 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26889 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92987 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7580 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81633 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49637 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7307 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/36042 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39893 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77643 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37054 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96812 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17174 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78261 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17430 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77451 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77475 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19138 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21933 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20517 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10368 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17184 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22509 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16925 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20377 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18708 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
```diff
@@ -231,6 +231,10 @@ enum class TapHandlingResult : uint8_t;
 #endif
 
+#if ENABLE(MODEL_PROCESS)
+- (void)willInvalidateDraggedModelWithContainerView:(UIView *)containerView;
```
You might want to prefix this internal method with an underscore, since `WKWebView` is public (we aren't super consistent about this rule, but it's generally a good idea to underscore-prefix internal methods on API objects to avoid conflicting with clients' category methods)

whsieh commented on Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.mm:
```diff
@@ -63,8 +64,46 @@
+    if (!m_modelPresentations.contains(layerIdentifier))
+        return nil;
+
+    ModelPresentation& modelPresentation = *(m_modelPresentations.get(layerIdentifier));
```
Nit - you can use `find` here to limit this to 1 hash lookup.

```suggestion
    auto iterator = m_modelPresentations.find(layerIdentifier);
    if (iterator == m_modelPresentations.end())
        return nil;

    auto& modelPresentation = iterator->value;
```

Or alternately, I think using `getOptional` would be both cleaner and safer from a smart pointer perspective:

```suggestion
    auto modelPresentation = m_modelPresentations.getOptional(layerIdentifier);
    // modelPresentation->…
```

achan00 replied:
Nice! Will use `getOptional`.

achan00 replied:
Actually need to do it the iterator way because the hash value type is a `UniqueRef` which can't create Optional with.

whsieh commented on Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.mm:
```diff
@@ -63,8 +64,46 @@
+    if (!m_modelPresentations.contains(layerIdentifier))
+        return;
+
+    ModelPresentation& modelPresentation = *(m_modelPresentations.get(layerIdentifier));
```
(Ditto)

whsieh commented on Source/WebKit/UIProcess/ios/DragDropInteractionState.mm:
```diff
@@ -302,14 +310,15 @@ inline static bool dragItemSupportsAsynchronousUpdates()
+            return createTargetedDragPreview((*image).get(), contentView, previewContainer, source.dragPreviewFrameInRootViewCoordinates, { }, nil, visiblePath, addPreviewViewToContainer).autorelease();
         }
-        return createTargetedDragPreview(source.image.get(), contentView, previewContainer, source.dragPreviewFrameInRootViewCoordinates, { }, nil, nil, addPreviewViewToContainer).autorelease();
+        return createTargetedDragPreview((*image).get(), contentView, previewContainer, source.dragPreviewFrameInRootViewCoordinates, { }, nil, nil, addPreviewViewToContainer).autorelease();
```
Nit - if possible, I think we'd prefer `image->get()` over `(*image).get()`

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -10268,6 +10271,18 @@ - (void)_startDrag:(RetainPtr<CGImageRef>)image item:(const WebCore::DragItem&)i
+    if (item.modelLayerID && _page) {
+        if (auto modelPresentationManager = _page->modelPresentationManagerProxy()) {
+            auto viewForDragPreview = modelPresentationManager->startDragForModel(*item.modelLayerID);
+            if (viewForDragPreview) {
```
```suggestion
            if (auto viewForDragPreview = modelPresentationManager->startDragForModel(*item.modelLayerID)) {
```

whsieh commented on Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
```diff
@@ -2242,6 +2237,57 @@ static BOOL isCompletelyWhite(UIImage *image)
+    auto messageHandler = adoptNS([[ModelLoadingMessageHandler alloc] init]);
+    [[configuration userContentController] addScriptMessageHandler:messageHandler.get() name:@"modelLoading"];
+
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 320, 500) configuration:configuration.get()]);
```
Nit - nowadays, we prefer using `RetainPtr` instead of `auto`, to make it clear where smart pointers have been deployed.

whsieh commented on Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
```diff
@@ -2242,6 +2237,57 @@ static BOOL isCompletelyWhite(UIImage *image)
+            [webView waitForNextPresentationUpdate];
+            NSArray *liftPreviews = [simulator liftPreviews];
+            UITargetedDragPreview *liftPreview = (UITargetedDragPreview *)liftPreviews.firstObject;
+            EXPECT_EQ(_UIRemoteView.class, liftPreview.view.class);
```
Maybe `-isKindOfClass:` instead? (So that it's robust in case it gets its own subclass in the future)

whsieh commented on Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
```diff
@@ -2242,6 +2237,57 @@ static BOOL isCompletelyWhite(UIImage *image)
+
+    NSArray *liftPreviews = [simulator liftPreviews];
+    EXPECT_EQ(1U, liftPreviews.count);
+    EXPECT_EQ(UITargetedDragPreview.class, [liftPreviews.firstObject class]);
```
Ditto

aprotyas commented on Source/WebKit/UIProcess/ios/DragDropInteractionState.mm:
```diff
@@ -412,11 +421,13 @@ inline static bool dragItemSupportsAsynchronousUpdates()
-                return adoptNS([[UIDragPreview alloc] initWithView:imageView.get() parameters:parameters.get()]).autorelease();
-            };
+            if (auto* draggedImage = std::get_if<RetainPtr<UIImage>>(&source.dragPreviewContent)) {
+                dragItem.previewProvider = [image = *draggedImage] () -> UIDragPreview * {
```
Style guide nit: no space between capture list and arg list. Also, we can omit empty arg lists in C++23!

```suggestion
                dragItem.previewProvider = [image = *draggedImage] -> UIDragPreview * {
```

aprotyas replied:
Is the return type hint needed?

achan00 replied:
Not needed. Removed.

aprotyas commented on Source/WebKit/UIProcess/ios/DragDropInteractionState.mm:
```diff
@@ -412,11 +421,13 @@ inline static bool dragItemSupportsAsynchronousUpdates()
+            if (auto* draggedImage = std::get_if<RetainPtr<UIImage>>(&source.dragPreviewContent)) {
+                dragItem.previewProvider = [image = *draggedImage] () -> UIDragPreview * {
+                    auto imageView = adoptNS([[UIImageView alloc] initWithImage:image.get()]);
+                    auto parameters = adoptNS([[UIDragPreviewParameters alloc] initWithTextLineRects:@[ [NSValue valueWithCGRect:[imageView bounds]] ]]);
```
We prefer explicit smart pointer type nowadays
```suggestion
                    RetainPtr imageView = adoptNS([[UIImageView alloc] initWithImage:image.get()]);
                    RetainPtr parameters = adoptNS([[UIDragPreviewParameters alloc] initWithTextLineRects:@[ [NSValue valueWithCGRect:[imageView bounds]] ]]);
```

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -10398,6 +10413,13 @@ - (void)cleanUpDragSourceSessionState
 
+#if ENABLE(MODEL_PROCESS)
+    if (_page) {
+        if (auto modelPresentationManager = _page->modelPresentationManagerProxy())
```
Ditto about explicit smart pointer preference
```suggestion
        if (RefPtr modelPresentationManager = _page->modelPresentationManagerProxy())
```

achan00 committed:
[visionOS] Fix the drag preview of the <model> element
https://bugs.webkit.org/show_bug.cgi?id=287525
rdar://141165213

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

For UITargetedDragPreview to capture the snapshot of a view
with entity-based content properly, we need to pass the _UIRemoteView
hosting the model content directly when constructing the UITargetedDragPreview.
To do that, we need to pass the model element’s PlatformLayerIdentifier
in the DragItem, so we can use it to retrieve the _UIRemoteView hosting
the model entity from ModelPresentationManagerProxy in the UI process
to stage the drag session with.
In DragDropInteractionState::createDragPreviewInternal(), if the drag
preview content is the _UIRemoteView rather than an image, create the
UITargetedDragPreview with that view directly.

Handle the edge case where the model element is removed while
it’s being dragged by adding it to the WKContentView’s drag previews
container view.

* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::doSystemDrag):
If the dragged element is an HTMLModelElement, save off its
PlatformLayerIdentifier in the DragItem.
* Source/WebCore/platform/DragItem.h:
Include the model element’s PlatformLayerIdentifier
in the DragItem struct.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _willInvalidateDraggedModelWithContainerView:]):
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.h:
* Source/WebKit/UIProcess/Model/ModelPresentationManagerProxy.mm:
(WebKit::ModelPresentationManagerProxy::startDragForModel):
Add the dragged model’s layer identifier to m_activelyDraggedModelLayerIDs.
Return the _UIRemoteView hosting the model entity.
(WebKit::ModelPresentationManagerProxy::doneWithCurrentDragSession):
Clear m_activelyDraggedModelLayerIDs.
(WebKit::ModelPresentationManagerProxy::invalidateModel):
If the model layer being removed is currently being dragged, call
-[WKWebView _willInvalidateDraggedModelWithContainerView:] with the
model container view so it can be re-parented to the window’s
view hierarchy.
* Source/WebKit/UIProcess/ios/DragDropInteractionState.h:
DragSourceState now stores content with the DragPreviewContentType
which is either a UIImage or a UIView.
* Source/WebKit/UIProcess/ios/DragDropInteractionState.mm:
(WebKit::shouldUseDragImageToCreatePreviewForDragSource):
(WebKit::DragDropInteractionState::previewForCancelling):
The UITargetedDragPreview’s view is the _UIRemoteView’s that’s
already in the remote layer tree’s view hierarchy if the dragged
element is a model element. Only add UITargetedDragPreview’s view
to m_previewViewsForDragCancel if its superview is the preview container.
(WebKit::DragDropInteractionState::createDragPreviewInternal const):
If the drag preview content is a UIView (which is the case for a
dragged model), create a UITargetedDragPreview with that UIView.
(WebKit::DragDropInteractionState::stageDragItem):
(WebKit::DragDropInteractionState::updatePreviewsForActiveDragSources):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _startDrag:item:]):
If the dragged item has a model layer identifier, stage the drag
item with the _UIRemoteView returned from the page’s ModelPresentationManagerProxy.
(-[WKContentView cleanUpDragSourceSessionState]):
Clear the dragged models tracked in ModelPresentationManagerProxy.
(-[WKContentView _willInvalidateDraggedModelWithContainerView:]):
This is called when the dragged model is being removed from the document.
We need to keep its container view in the window by hiding it and
adding it to the WKContentView’s _dragPreviewContainerView.
* Source/WebKit/UIProcess/ios/WKPageHostedModelView.mm:
(-[WKPageHostedModelView init]):
Add some nil checks which are needed for TestWebKitAPI model drag
and drop tests.
* Tools/TestRunnerShared/spi/UIKitSPIForTesting.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/simple-model-page.html: Added.
Simple test page with a model. Posts messages when the model is
loaded and when it’s ready.
* Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm:
(-[ModelLoadingMessageHandler userContentController:didReceiveScriptMessage:]):
Teach ModelLoadingMessageHandler to catch both the model loaded
event and the model ready event.
(TestWebKitAPI::TEST(DragAndDropTests, CanStartDragOnModel)):
Update test to load the new simple-model-page.html.
(TestWebKitAPI::TEST(DragAndDropTests, CheckModelDragPreview)):
Simulate the dragging of a model element and verify that the resulting
UITargetedDragPreview’s view is a _UIRemoteView. In the middle of the
drag, remove the model from the document and make sure the _UIRemoteView
is still in the window’s view hierarchy. We have to test this scenario
while the drag is still in progress because the page’s drag preview
container view gets removed from the window at the end of the drag session.
* Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:
* Tools/TestWebKitAPI/ios/DragAndDropSimulatorIOS.mm:
(-[DragAndDropSimulator _advanceProgress]):
Call _dragProgressMidPointReachedBlock if it's set when the
drag progress has passed the midway point.

Canonical link: https://commits.webkit.org/290437@main
